File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int count = 0;
    int ans = Integer.MAX_VALUE;

    public void mins(TreeNode root) {
        if (root == null) {
            return;
        }
        count++;
        if (root.left==null && root.right==null) {
            ans=Math.min(ans, count);
        }
        mins(root.left);
        mins(root.right);
        count--;
    }

    public int minDepth(TreeNode root) {
        if(root==null) {
            return 0;
        }
        mins(root);
        return ans;
    }
}

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solver(TreeNode*root){
        if(!root) return 0;
        if(!root->left && !root->right){
            return 1;
        }
        int lh = 1 + solver(root->left);
        int rh = 1 +  solver(root->right);
        if(lh==1 ){
            return rh;
        }
        if(rh==1 ){
            return lh;
        }
        
        return min(lh,rh);
    }
    int minDepth(TreeNode* root) {
        if(!root) return 0;
        return solver(root);
    }
};